### PR TITLE
Update hessian matrix code to include order kwarg

### PR DESCRIPTION
**Pull request by JohnnyTeutonic**
## Description

I've included an optional parameter into the Hessian matrix function in corner.py that allows for reverse and forward ordering of the Hessian matrix. This 'order' parameter defaults to Fortran order if the order isn't specificied, wherein, Hxx and Hyy will be reversed if the image dimensions are 2, and otherwise will default to C style.
I also changed the corresponding test function 'test_hessian_matrix_eigvals' in test_corner.py to 'force' the order to 'C', and likewise did the same in _frangi.py, line 47. 
The latter in _frangi.py:
# Make 2D hessian

(Dyy, Dxy, Dxx) = hessian_matrix(image, sigma, order='C')
This is my first pull request so please let me know what I can do if this is not sufficient.
Thanks!
## Checklist

[It's fine to submit PRs which are a work in progress! But before they are merged, all PRs should provide:]
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Unit tests

[For detailed information on these and other aspects see [scikit-image contribution guidelines](http://scikit-image.org/docs/dev/contribute.html)]
## References

Closes issue #2301

### CODE REVIEW

**Comment by JohnnyTeutonic** [comment]
Sorry for the 'git' noise! :s

**Comment by sciunto** [comment]
I assign to @ahojnnes and @jni who already worked on this feature.

File: skimage/feature/corner.py [hunk 1]
```diff
@@ -122,9 +122,13 @@ def hessian_matrix(image, sigma=1, mode='constant', cval=0):
         weighting function for the auto-correlation matrix.
     mode : {'constant', 'reflect', 'wrap', 'nearest', 'mirror'}, optional
         How to handle values outside the image borders.
-    cval : float, optional
+    cval : float, 'F'
```
Comment (ahojnnes) [inline skimage/feature/corner.py]: This is probably not intended.

File: skimage/feature/corner.py [hunk 2]
```diff
@@ -159,10 +163,18 @@ def hessian_matrix(image, sigma=1, mode='constant', cval=0):
     H_elems = [np.gradient(gradients[ax0], axis=ax1)
                for ax0, ax1 in combinations_with_replacement(axes, 2)]
 
-    if image.ndim == 2:
-        # The legacy 2D code followed (x, y) convention, so we swap the axis
-        # order to maintain compatibility with old code
+
+    if order is None:
```
Comment (ahojnnes) [inline skimage/feature/corner.py]: We should raise a deprecation warning here.

**Comment by JohnnyTeutonic** [comment]
Okay I can put in a depreciation warning there sure

**Comment by JohnnyTeutonic** [comment]
Deprecation*

**Comment by jni** [comment]
Hey @JohnnyTeutonic, here's an example of how we do deprecation of default values:

https://github.com/scikit-image/scikit-image/blob/master/skimage/util/dtype.py#L51

(Don't forget to `from warnings import warn` at the top of the file.)

Also, in `TODO.txt`, add a note to change this default value in the 0.15 section.

**Comment by JohnnyTeutonic** [comment]
Okay thanks Juan

**Comment by JohnnyTeutonic** [comment]
Didn't see that, will fix

File: skimage/feature/corner.py [hunk 3]
```diff
@@ -122,9 +122,13 @@ def hessian_matrix(image, sigma=1, mode='constant', cval=0):
         weighting function for the auto-correlation matrix.
     mode : {'constant', 'reflect', 'wrap', 'nearest', 'mirror'}, optional
         How to handle values outside the image borders.
-    cval : float, optional
+    cval : float, 'F'
+        else:optional
         Used in conjunction with mode 'constant', the value outside
         the image boundaries.
+    order : {'C', 'F'}, optional
+        this parameter allows for the use of reverse or forward order of
```
Comment (jni) [inline skimage/feature/corner.py]: this -> This. Also, no need to specify "this parameter". How about:

It just occurred to me, that, even though in the discussion in #2301 we settled on C|F, actually it's not semantically correct: the order is reversed, not simply transposed. So I suggest changing this to `order : {'xy', 'rc'}`. @ahojnnes thoughts?

File: skimage/feature/tests/test_corner.py [hunk 4]
```diff
@@ -41,7 +41,7 @@ def test_structure_tensor():
 def test_hessian_matrix():
     square = np.zeros((5, 5))
     square[2, 2] = 4
-    Hxx, Hxy, Hyy = hessian_matrix(square, sigma=0.1)
+    Hyy, Hxy, Hxx = hessian_matrix(square, sigma=0.1, order='C')
```
Comment (jni) [inline skimage/feature/tests/test_corner.py]: I suggest renaming these to `Hrr, Hrc, Hcc`.

File: skimage/feature/tests/test_corner.py [hunk 5]
```diff
@@ -94,8 +94,8 @@ def test_structure_tensor_eigvals():
 def test_hessian_matrix_eigvals():
     square = np.zeros((5, 5))
     square[2, 2] = 4
-    Hxx, Hxy, Hyy = hessian_matrix(square, sigma=0.1)
-    l1, l2 = hessian_matrix_eigvals(Hxx, Hxy, Hyy)
+    Hyy, Hxy, Hxx = hessian_matrix(square, sigma=0.1, order='C')
+    l1, l2 = hessian_matrix_eigvals(Hyy, Hxy, Hxx)
```
Comment (jni) [inline skimage/feature/tests/test_corner.py]: rename as above

File: skimage/filters/_frangi.py [hunk 6]
```diff
@@ -44,15 +44,15 @@ def _frangi_hessian_common_filter(image, scale_range, scale_step,
     # Filtering for all sigmas
     for i, sigma in enumerate(sigmas):
         # Make 2D hessian
-        (Dxx, Dxy, Dyy) = hessian_matrix(image, sigma)
+        (Dyy, Dxy, Dxx) = hessian_matrix(image, sigma, order='C')
```
Comment (jni) [inline skimage/filters/_frangi.py]: Rename here also.

**Comment by ahojnnes** [comment]
@JohnnyTeutonic I just merged a new PR #2312 which uses the Hessian matrix code. You should rebase your PR here and update the code there accordingly to hide the deprecation warning. The procedure is described here: https://github.com/scikit-image/scikit-image/blob/master/.github/CONTRIBUTING.txt#L101

**Comment by JohnnyTeutonic** [comment]
alright, @jni @ahojnnes, I shall rebase the PR and am working on updating the code to hide the deprecation warning. Might take me some time but will let you know when I make some progress.

**Comment by jni** [comment]
@JohnnyTeutonic looks like you've messed up your rebase quite dramatically... =P Let's get together some time tomorrow to discuss.

**Comment by JohnnyTeutonic** [comment]
Not sure how I've messed it up? I did a push onto my remote Hessian branch
to update it from my local Hessian branch and I looked at the code and it
seemed fine to me. The master branch is intact, as I didn't merge onto it
unlike last time.
But I can meet tomorrow. When is good for you? 12?

**Comment by jni** [comment]
@JohnnyTeutonic you're passing order='C' instead of order='rc'. ;)

btw you can check which tests are failing in a PR by clicking on "details" on the Travis line, and then clicking on one of the results. (I usually go for 3.5 ;). You'll get a long output, and towards the end you'll see a list of the failing tests.

**Comment by JohnnyTeutonic** [comment]
Okay thanks will look into it!

File: skimage/feature/corner.py [hunk 7]
```diff
@@ -123,25 +123,29 @@ def hessian_matrix(image, sigma=1, mode='constant', cval=0):
     mode : {'constant', 'reflect', 'wrap', 'nearest', 'mirror'}, optional
         How to handle values outside the image borders.
     cval : float, optional
+        else:optional
```
Comment (jni) [inline skimage/feature/corner.py]: errr, what's going on here?

File: skimage/feature/corner.py [hunk 8]
```diff
@@ -123,25 +123,29 @@ def hessian_matrix(image, sigma=1, mode='constant', cval=0):
...
         Used in conjunction with mode 'constant', the value outside
         the image boundaries.
+    order : {'xy', 'rc'}, optional
+        this parameter allows for the use of reverse or forward order of
+        returned matrix values
```
Comment (jni) [inline skimage/feature/corner.py]: Add proper capitalisation and full stop. Also, I wouldn't call this reverse order of matrix values, but rather, reverse order of image axes in gradient computation. 'rc' indicates first axis first (Hrr), while 'xy' indicates last axis first (Hxx, which is the same as Hcc in 2D). Try to get this across in the docstring (feel free to reuse my text).

File: skimage/feature/corner.py [hunk 9]
```diff
@@ -123,25 +123,29 @@ def hessian_matrix(image, sigma=1, mode='constant', cval=0):
...
         Element of the Hessian matrix for each pixel in the input image.
 
     Examples
     --------
     >>> from skimage.feature import hessian_matrix
     >>> square = np.zeros((5, 5))
     >>> square[2, 2] = 4
-    >>> Hxx, Hxy, Hyy = hessian_matrix(square, sigma=0.1)
-    >>> Hxy
+    >>> Hrr, Hrc, Hcc = hessian_matrix(square, sigma=0.1)
```
Comment (jni) [inline skimage/feature/corner.py]: This docstring example is incorrect, because without order='rc', it will return Hxx, Hxy, Hyy. So I would include order='rc' in the docstring call.

File: skimage/feature/corner.py [hunk 10]
```diff
@@ -154,18 +158,29 @@ def hessian_matrix(image, sigma=1, mode='constant', cval=0):
     gaussian_filtered = ndi.gaussian_filter(image, sigma=sigma,
                                             mode=mode, cval=cval)
 
+    if order is None:
+        if image.ndim == 2:
+            # The legacy 2D code followed (x, y) convention, so we swap the axis
+            # order to maintain compatibility with old code
+            warn('deprecation warning: the default order of the hessian matrix values '
+            'will be ''row-column'' instead of ''xy'' starting in skimage version 0.15'
+             'Use order=''rc'' or order=''xy'' to set this explicitly')
```
Comment (jni) [inline skimage/feature/corner.py]: Use consistent indentation here — the quotes should be aligned with the open parenthesis. Also, this ain't LaTeX =P, so use normal double quotes instead of two single quotes. (In fact I just tested it and two single quotes simply disappear, as they count as closing and re-opening the string.)

File: skimage/feature/tests/test_corner.py [hunk 11]
```diff
@@ -41,30 +41,37 @@ def test_structure_tensor():
...
 
-    assert_almost_equal(Hyy, np.array([[0, 0,  2, 0, 0],
+    assert_almost_equal(Hcc, np.array([[0, 0,  2, 0, 0],
                                        [0, 0,  0, 0, 0],
                                        [0, 0, -2, 0, 0],
                                        [0, 0,  0, 0, 0],
                                        [0, 0,  2, 0, 0]]))
 
+    matrix2d = np.random.rand(3,3)
+    Arr, Arc, Acc = hessian_matrix(matrix2d, sigma=0.1)
```
Comment (jni) [inline skimage/feature/tests/test_corner.py]: I suspect that this call is what's causing Python 2 not to warn — the warning might only be raised once in Py2 for some reason. Since this line is doing nothing useful, just remove it — the `assert_warns` function will call `hessian_matrix` as appropriate.

File: skimage/feature/corner.py (hunk 9 above)
Comment (jni) [inline skimage/feature/corner.py]: @JohnnyTeutonic this still needs updating... *and* is the reason for the failing test in 2.7: the warning is raised here and so it doesn't get raised again in the tests. (See line 1800 of the build: https://travis-ci.org/scikit-image/scikit-image/jobs/173806370#L1800)

File: skimage/feature/corner.py (hunk 10 above)
Comment (jni) [inline skimage/feature/corner.py]: @JohnnyTeutonic this indentation still needs fixing

File: skimage/feature/corner.py [hunk 12]
```diff
@@ -320,7 +336,8 @@ def shape_index(image, sigma=1, mode='constant', cval=0):
         Standard deviation used for the Gaussian kernel, which is used for
         smoothing the input data before Hessian eigen value calculation.
     mode : {'constant', 'reflect', 'wrap', 'nearest', 'mirror'}, optional
-        How to handle values outside the image borders.
+        How to handle values outside the image borders.from .._shared.utils import skimage_deprecation, warn
```
Comment (jni) [inline skimage/feature/corner.py]: @JohnnyTeutonic another stray paste...

File: skimage/feature/tests/test_corner.py [hunk 13]
```diff
@@ -41,30 +41,36 @@ def test_structure_tensor():
...
+
+
+
+
 
 def test_hessian_matrix_3d():
     cube = np.zeros((5, 5, 5))
     cube[2, 2, 2] = 4
-    Hs = hessian_matrix(cube, sigma=0.1)
+    Hs = hessian_matrix(cube, sigma=0.1, order=None)
```
Comment (jni) [inline skimage/feature/tests/test_corner.py]: Why order=None here? Use `order='rc'` except when we want the warning to happen.

File: skimage/feature/corner.py [hunk 14]
```diff
@@ -154,18 +159,29 @@ def hessian_matrix(image, sigma=1, mode='constant', cval=0):
...
                                             mode=mode, cval=cval)
 
+    if order is None:
+        if image.ndim == 2:
+            # The legacy 2D code followed (x, y) convention, so we swap the axis
+            # order to maintain compatibility with old code
+            warn('deprecation warning: the default order of the hessian matrix values '
+            'will be "row-column" instead of "xy" starting in skimage version 0.15'
+             'Use order="rc" or order="xy" to set this explicitly')
+            order = 'rc'
```
Comment (jni) [inline skimage/feature/corner.py]: Oh LOL this is actually backwards: default for None/2D is xy, otherwise it's rc.

File: skimage/feature/corner.py [hunk 15]
```diff
@@ -320,7 +336,7 @@ def shape_index(image, sigma=1, mode='constant', cval=0):
         Standard deviation used for the Gaussian kernel, which is used for
         smoothing the input data before Hessian eigen value calculation.
     mode : {'constant', 'reflect', 'wrap', 'nearest', 'mirror'}, optional
-        How to handle values outside the image borders.
+        How to handle values outside the image borders
```
Comment (jni) [inline skimage/feature/corner.py]: Undo this change. (add back the full stop)

**Comment by JohnnyTeutonic** [comment]
I thought default was xy, was very confused by your request :p

File: skimage/feature/corner.py (hunk 14 above)
Comment (JohnnyTeutonic) [inline skimage/feature/corner.py]: oh, wait. do you mean I should have: 

    if order is None:
        if image.ndim == 2:
            # The legacy 2D code followed (x, y) convention, so we swap the axis
            # order to maintain compatibility with old code
            warn('deprecation warning: the default order of the hessian matrix values '
            'will be "row-column" instead of "xy" starting in skimage version 0.15'
             'Use order="rc" or order="xy" to set this explicitly')
            order = 'xy'
        else:
            order = 'rc'



    if order == 'rc':
        axes = reversed(axes)

??

Comment (JohnnyTeutonic) [inline skimage/feature/corner.py]: Sorry, i meant:

if order is None:
    if image.ndim == 2:
        # The legacy 2D code followed (x, y) convention, so we swap the axis
        # order to maintain compatibility with old code
        warn('deprecation warning: the default order of the hessian matrix values '
        'will be "row-column" instead of "xy" starting in skimage version 0.15'
         'Use order="rc" or order="xy" to set this explicitly')
        order = 'rc'
    else:
        order = 'xy'



if order == 'rc':
    axes = reversed(axes)

Comment (JohnnyTeutonic) [inline skimage/feature/corner.py]: @jni  Sorry, my first post was what i meant...

**Comment by jni** [comment]
@JohnnyTeutonic 
1. your latest commit message and actual changes don't match:
https://github.com/scikit-image/scikit-image/pull/2327/commits/ee112e3b90793edc95aba3404b0d56b80fa3f1dc
2. In those changes, the alignment is still wrong: the subsequent lines should line up with the open parenthesis, like so:

```python
warn('hello '
     'world')
```

**Comment by JohnnyTeutonic** [comment]
Sorry, what was wrong with my commit message?

**Comment by JohnnyTeutonic** [comment]
@jni. I still can't figure out why the code is still failing under python2.7 but not in python3. Really stumped about this. 
Do you think this code is okay here? (from _frangi.py):

line 45: # Filtering for all sigmas
    for i, sigma in enumerate(sigmas):
        # Make 2D hessian
        (Drr, Drc, Dcc) = hessian_matrix(image, sigma, order='rc')

        # Correct for scale
        Drr = (sigma ** 2) * Drr
        Drc = (sigma ** 2) * Drc
        Dcc = (sigma ** 2) * Dcc

        # Calculate (abs sorted) eigenvalues and vectors
        (lambda1, lambda2) = hessian_matrix_eigvals(Drr, Drc, Dcc)

We're 'forcing' the order to 'rc' here, and as a result the order of the axes will change. However, I've used the notation of 'Drr' instead of 'Dyy' here.

**Comment by jni** [comment]
@JohnnyTeutonic look at the commit I linked:
https://github.com/scikit-image/scikit-image/pull/2327/commits/ee112e3b90793edc95aba3404b0d56b80fa3f1dc
You can see both the changes and the commit message. You can see that they have nothing at all to do with one another.

The code change you pointed to is fine. But there is a another doctest that is failing. In the link, you can tell where in the source code there is a warning being raised:
https://travis-ci.org/scikit-image/scikit-image/jobs/174091286#L1801
Because the warning is raised earlier, it doesn't get raised again when you actually test it. So you need to change that call to use order='rc'.

**Comment by JohnnyTeutonic** [comment]
@jni. Well thanks for pointing out where in the build i can find the error - I didn't realise that you could determine where in the code the errors derive from. 
As for your comment re: commit no. ee112e3 about my commit message not being related to my commit, yes, I realise now that I used the word 'doctest' inappropriately...

**Comment by jni** [comment]
Whoo! =)

**Comment by JohnnyTeutonic** [comment]
About time!
